src/compile/data.ts

```diff
@@ -23,79 +23,228 @@ const DEFAULT_NULL_FILTERS = {
 };
 
 /**
- * Create Vega's data array from a given model.
+ * Composable component instance of a model's data.
+ */
+export interface DataComponent {
+  source?: VgData;
+  formatParse?: Dict<string>;
+
+  /** String set of fields for null filtering */
+  nullFilter?: StringSet;
+
+  /** Hashset of a formula object */
+  calculate?: Dict<Formula>;
+
+  /** Filter test expression */
+  filter?: string;
+
+  /** Dictionary mapping a bin parameter hash to transforms of the binned field */
+  bin?: Dict<VgTransform[]>;
+
+  /** Dictionary mapping an output field name (hash) to the time unit transform  */
+  timeUnit?: Dict<VgTransform>;
+
+  /** String set of fields to be filtered */
+  nonPositiveFilter?: StringSet;
+
+  /** Data source for feeding stacked scale. */
+  // TODO: need to revise if single VgData is sufficient with layer / concat
+  stackScale?: VgData;
+
+  /** Dictionary mapping an output field name (hash) to the sort and rank transforms  */
+  colorRank?: Dict<VgTransform[]>;
+
+  /** String set of time units that need their own data sources for scale domain */
+  timeUnitDomain?: StringSet;
+
+  /** Array of summary component object for producing summary (aggregate) data source */
+  summary?: SummaryComponent[];
+}
+
+/**
+ * Composable component for a model's summary data
+ */
+interface SummaryComponent {
+  /** Name of the summary data source */
+  name: string;
+
+  /** Source for the summary data source */
+  source: string;
+
+  /** String set for all dimension fields  */
+  dimensions: StringSet;
+
+  /** dictionary mapping field name to string set of aggregate ops */
+  measures: Dict<StringSet>;
+}
+
+// TODO: split this file into multiple files and remove this linter flag
+/* tslint:disable:no-use-before-declare */
+
+export function compileUnitData(model: UnitModel): DataComponent {
```

Comment:
Note: I'm not using for loop for `compileUnitData` and `compileFacetData` here like in axis/scale/legend so that we can ensure type check.